File: src-python/infra/vector_index/chunker.py
"""章节文本切块。参见 PRD §5.2。

切分规则：
- frontmatter 剥离（python-frontmatter，不用正则）
- 按段落切（空行或 ## 标题为边界）
- 切分点在句号/叹号/问号处
- < 100 字合并到相邻段
- > 600 字先按句号切分再组合
- Overlap 用"最后一整句"
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any


@dataclass
class ChunkData:
    """切块结果（内部使用，不依赖 domain Chunk 避免循环导入）。"""

    content: str
    chapter_num: int
    chunk_index: int
    branch_id: str = "main"
    characters: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)


# 句子结束标点
_SENTENCE_END = re.compile(r"[。！？…\n]")


def split_chapter_into_chunks(
    text: str,
    chapter_num: int,
    max_size: int = 500,
    overlap_sentences: int = 1,
) -> list[ChunkData]:
    """将章节文本切块（PRD §5.2）。

    Args:
        text: 章节原始文本（可能含 frontmatter）。
        chapter_num: 章节号。
        max_size: 每个 chunk 最大字符数。
        overlap_sentences: overlap 句数。

    Returns:
        ChunkData 列表。
    """
    # 第零步：剥离 frontmatter
    import frontmatter as fm
    post = fm.loads(text)
    body: str = post.content.strip()

    if not body:
        return []

    # 按段落切（空行或 ## 标题为边界）
    raw_paragraphs = re.split(r"\n\s*\n|(?=^##\s)", body, flags=re.MULTILINE)
    paragraphs = [p.strip() for p in raw_paragraphs if p.strip()]

    # < 100 字合并到相邻段
    merged = _merge_short_paragraphs(paragraphs, min_size=100)

    # > 600 字按句号切分再组合
    expanded: list[str] = []
    for para in merged:
        if len(para) > 600:
            expanded.extend(_split_long_paragraph(para, max_size))
        else:
            expanded.append(para)

    # 按 max_size 组合成 chunks
    chunks_text = _combine_into_chunks(expanded, max_size)

    # 添加 overlap（最后一整句）
    if overlap_sentences > 0 and len(chunks_text) > 1:
        chunks_text = _add_overlap(chunks_text, overlap_sentences)

    # 构建 ChunkData
    result: list[ChunkData] = []
    for i, content in enumerate(chunks_text):
        result.append(ChunkData(
            content=content,
            chapter_num=chapter_num,
            chunk_index=i,
            branch_id="main",
        ))

    return result


def _merge_short_paragraphs(paragraphs: list[str], min_size: int) -> list[str]:
    """< min_size 字的段落合并到相邻段。"""
    if not paragraphs:
        return []
    merged: list[str] = [paragraphs[0]]
    for para in paragraphs[1:]:
        if len(merged[-1]) < min_size:
            merged[-1] += "\n" + para
        else:
            merged.append(para)
    # 最后一段也可能太短
    if len(merged) > 1 and len(merged[-1]) < min_size:
        merged[-2] += "\n" + merged[-1]
        merged.pop()
    return merged


def _split_long_paragraph(para: str, max_size: int) -> list[str]:
    """按句号切分长段落，再组合成 <= max_size 的块。"""
    sentences = _split_sentences(para)
    result: list[str] = []
    current = ""
    for sent in sentences:
        if current and len(current) + len(sent) > max_size:
            result.append(current.strip())
            current = sent
        else:
            current += sent
    if current.strip():
        result.append(current.strip())
    return result if result else [para]


def _split_sentences(text: str) -> list[str]:
    """按句子结束标点切分。保留标点在句子末尾。"""
    parts: list[str] = []
    last = 0
    for m in _SENTENCE_END.finditer(text):
        end = m.end()
        parts.append(text[last:end])
        last = end
    if last < len(text):
        parts.append(text[last:])
    return [p for p in parts if p.strip()]


def _combine_into_chunks(paragraphs: list[str], max_size: int) -> list[str]:
    """将段落组合成 <= max_size 的 chunks。"""
    if not paragraphs:
        return []
    chunks: list[str] = [paragraphs[0]]
    for para in paragraphs[1:]:
        if len(chunks[-1]) + len(para) + 1 <= max_size:
            chunks[-1] += "\n" + para
        else:
            chunks.append(para)
    return chunks


def _add_overlap(chunks: list[str], n_sentences: int) -> list[str]:
    """为相邻 chunk 添加 overlap（最后 n 整句）。"""
    result = [chunks[0]]
    for i in range(1, len(chunks)):
        prev_sentences = _split_sentences(chunks[i - 1])
        overlap = "".join(prev_sentences[-n_sentences:]) if prev_sentences else ""
        if overlap:
            result.append(overlap + chunks[i])
        else:
            result.append(chunks[i])
    return result

File: src-engine/vector/chunker.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.

/**
 * 章节文本切块。参见 PRD §5.2。
 *
 * 切分规则：
 * - frontmatter 剥离（safeMatter，仅剥真章节 frontmatter）
 * - 按段落切（空行或 ## 标题为边界）
 * - 切分点在句号/叹号/问号处
 * - < 100 字合并到相邻段
 * - > 600 字先按句号切分再组合
 * - Overlap 用"最后一整句"
 */

import { KNOWN_CHAPTER_META_KEYS } from "../domain/chapter.js";
import { safeMatter } from "../domain/frontmatter.js";
import { scanCharactersInChapter } from "../domain/character_scanner.js";

/** 切块结果。 */
export interface ChunkData {
  content: string;
  chapter_num: number;
  chunk_index: number;
  branch_id: string;
  characters: string[];
  metadata: Record<string, unknown>;
}

/** 角色名册（与 Project.cast_registry 兼容）。 */
export interface CastRegistryLike {
  characters?: string[];
}

/** 句子结束标点。 */
const SENTENCE_END = /[。！？…\n]/g;

/**
 * 单块「出场角色」标签判据（TD-020：与状态扫描同源 scanCharactersInChapter，
 * 供表后通篇只用别名的文字段也能记到主名）。存量库的 metadata 重扫
 * （RagManager.rescanChunkCharacters）与新建块共用此判据，禁两处各写。
 */
export function scanChunkCharacters(
  content: string,
  cast_registry: CastRegistryLike | null | undefined,
  character_aliases: Record<string, string[]> | null | undefined,
  chapter_num: number,
): string[] {
  if (!cast_registry?.characters?.length) return [];
  return Object.keys(scanCharactersInChapter(content, cast_registry, character_aliases ?? null, chapter_num));
}

/**
 * 将章节文本切块（PRD §5.2）。
 */
export function splitChapterIntoChunks(
  text: string,
  chapter_num: number,
  max_size = 500,
  overlap_sentences = 1,
  cast_registry?: CastRegistryLike | null,
  character_aliases?: Record<string, string[]> | null,
): ChunkData[] {
  // 剥离 frontmatter（审计 B-2）：输入通常已是 content-only 正文（rag_manager 传
  // get_content_only 结果），这里的剥离只是对「误传整文件」的防御。裸 matter()
  // 会把 `---` 开头的首场景吞成 frontmatter（该场景永不进向量索引），非法 YAML
  // 形态则直接抛错（indexChapter 失败）—— safeMatter 只在有真章节 frontmatter
  // 时才剥，其余整文进分块。
  const parsed = safeMatter(text, KNOWN_CHAPTER_META_KEYS);
  const body = parsed.content.trim();

  if (!body) return [];

  // 按段落切（空行或 ## 标题为边界）
  const rawParagraphs = body.split(/\n\s*\n|(?=^##\s)/m);
  const paragraphs = rawParagraphs.map((p) => p.trim()).filter(Boolean);

  // < 100 字合并到相邻段
  const merged = mergeShortParagraphs(paragraphs, 100);

  // > 600 字按句号切分再组合
  const expanded: string[] = [];
  for (const para of merged) {
    if (para.length > 600) {
      expanded.push(...splitLongParagraph(para, max_size));
    } else {
      expanded.push(para);
    }
  }

  // 按 max_size 组合成 chunks
  let chunksText = combineIntoChunks(expanded, max_size);

  // 添加 overlap
  if (overlap_sentences > 0 && chunksText.length > 1) {
    chunksText = addOverlap(chunksText, overlap_sentences);
  }

  // 构建 ChunkData（含逐块角色扫描；判据单源 scanChunkCharacters，认别名）
  return chunksText.map((content, i) => {
    const characters = scanChunkCharacters(content, cast_registry, character_aliases, chapter_num);
    return {
      content,
      chapter_num,
      chunk_index: i,
      branch_id: "main",
      characters,
      metadata: {},
    };
  });
}

function mergeShortParagraphs(paragraphs: string[], minSize: number): string[] {
  if (paragraphs.length === 0) return [];
  const merged = [paragraphs[0]];
  for (let i = 1; i < paragraphs.length; i++) {
    if (merged[merged.length - 1].length < minSize) {
      merged[merged.length - 1] += "\n" + paragraphs[i];
    } else {
      merged.push(paragraphs[i]);
    }
  }
  // 最后一段也可能太短
  if (merged.length > 1 && merged[merged.length - 1].length < minSize) {
    merged[merged.length - 2] += "\n" + merged[merged.length - 1];
    merged.pop();
  }
  return merged;
}

function splitLongParagraph(para: string, maxSize: number): string[] {
  const sentences = splitSentences(para);
  const result: string[] = [];
  let current = "";
  for (const sent of sentences) {
    if (current && current.length + sent.length > maxSize) {
      result.push(current.trim());
      current = sent;
    } else {
      current += sent;
    }
  }
  if (current.trim()) {
    result.push(current.trim());
  }
  return result.length > 0 ? result : [para];
}

function splitSentences(text: string): string[] {
  const parts: string[] = [];
  let last = 0;
  const re = new RegExp(SENTENCE_END.source, "g");
  for (const m of text.matchAll(re)) {
    parts.push(text.slice(last, m.index + m[0].length));
    last = m.index + m[0].length;
  }
  if (last < text.length) {
    parts.push(text.slice(last));
  }
  return parts.filter((p) => p.trim());
}

function combineIntoChunks(paragraphs: string[], maxSize: number): string[] {
  if (paragraphs.length === 0) return [];
  const chunks = [paragraphs[0]];
  for (let i = 1; i < paragraphs.length; i++) {
    if (chunks[chunks.length - 1].length + paragraphs[i].length + 1 <= maxSize) {
      chunks[chunks.length - 1] += "\n" + paragraphs[i];
    } else {
      chunks.push(paragraphs[i]);
    }
  }
  return chunks;
}

function addOverlap(chunks: string[], nSentences: number): string[] {
  const result = [chunks[0]];
  for (let i = 1; i < chunks.length; i++) {
    const prevSentences = splitSentences(chunks[i - 1]);
    const overlap = prevSentences.slice(-nSentences).join("");
    result.push(overlap ? overlap + chunks[i] : chunks[i]);
  }
  return result;
}
